x/wasm/ibc.go

```diff
@@ -280,10 +280,16 @@ func (i IBCHandler) OnRecvPacket(
 	if err != nil {
 		ack = channeltypes.NewErrorAcknowledgement(err)
 		// the state gets reverted, so we drop all captured events
-	} else if ack == nil || ack.Success() {
+	} else if ack == nil || ack.Acknowledgement() == nil {
 		// emit all contract and submessage events on success
 		// nil ack is a success case, see: https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/keeper/msg_server.go#L453
 		ctx.EventManager().EmitEvents(em.Events())
+		// contract wants async acknowledgement, so store the packet for later
+		i.keeper.StoreAsyncAckPacket(ctx, packet)
+		ack = nil
+	} else if ack.Success() {
+		// emit all contract and submessage events on success
+		ctx.EventManager().EmitEvents(em.Events())
```

Comment:
This might be easier to read if it is not chained after the error handling, for example something like:
```suggestion
        types.EmitAcknowledgementEvent(ctx, contractAddr, ack, err)
        return ack
	}
	
	switch {
	case ack == nil || ack.Acknowledgement() == nil:
		// emit all contract and submessage events on success
		// nil ack is a success case, see: https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/keeper/msg_server.go#L453
		ctx.EventManager().EmitEvents(em.Events())
		// contract wants async acknowledgement, so store the packet for later
		i.keeper.StoreAsyncAckPacket(ctx, packet)
		ack = nil
	case ack.Success():
		// emit all contract and submessage events on success
		ctx.EventManager().EmitEvents(em.Events())
    default:
        panic("unreachable")
    }
```
it also has the added benefit of panicing in an unexpected scenario

Reply:
I like the early return and the switch, but that default case is actually reachable. That `keeper.OnRecvPacket` call translates contract errors to error ack already. Tbh, this whole split between `IBCHandler.OnRecvPacket` and `Keeper.OnRecvPacket` seems very confusing to me.